File: attic/lighthouse_ros/lighthouse_ros/lighthouse_protocol_decoder.py
# Copyright 2024 Ekumen, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from lighthouse_ros.frame_decoders import SyncFrameDecoder, DataFrameDecoder
from lighthouse_ros.ootx_frame_decoder import OOTXFrameDecoder
from lighthouse_ros.sweep_processor import SweepProcessor
from lighthouse_ros.measurement_processor import MeasurementProcessor
from lighthouse_ros.types import DataFrameContents
from lighthouse_ros.types import SweepBlockRawData
from lighthouse_ros.types import SweepBlockBearings
from lighthouse_ros.utils import DECK_LIGHTHOUSE_MAX_N_BS
from lighthouse_ros.utils import MIN_TICKS_BETWEEN_SLOW_BITS
from lighthouse_ros.utils import PULSE_PROCESSOR_N_SENSORS
from lighthouse_ros.utils import timestamp_diff
from lighthouse_ros.utils import timestamp_abs_diff_larger_than
from lighthouse_ros.logger_proxy import LoggerProxy
from lighthouse_ros.types import BearingCallback

from typing import Optional


class LighthouseProtocolDecoder:
    """Decoder for the lighthouse."""

    MODE_SYNC: int = 0
    MODE_DATA: int = 1

    def __init__(
        self,
        logger: LoggerProxy,
        app_bearing_callback: Optional[BearingCallback] = None,
    ):
        """Initialize the decoder."""
        self.__logger = logger
        self.__app_bearing_callback = app_bearing_callback
        self.current_mode = self.MODE_SYNC
        self.sync_frame_decoder = SyncFrameDecoder(
            callback=self.sync_frame_detected_callback,
            logger=self.__logger,
        )
        self.data_frame_decoder = None
        self.sweep_decoder = SweepProcessor(
            callback=self.sweep_callback,
            logger=self.__logger,
        )
        self.measurement_processor = MeasurementProcessor(
            callback=self.measurement_callback,
            logger=self.__logger,
        )
        self.ootx_decoders = [None for _ in range(DECK_LIGHTHOUSE_MAX_N_BS)]
        self.prev_timestamp0 = [0 for _ in range(DECK_LIGHTHOUSE_MAX_N_BS)]

    def process_byte(self, byte: int) -> None:
        """Process a byte."""
        try:
            if self.current_mode == self.MODE_SYNC:
                self.sync_frame_decoder.process_byte(byte)
            elif self.current_mode == self.MODE_DATA:
                self.data_frame_decoder.process_byte(byte)
        except Exception as e:
            self.__logger.error(f"Exception caught: {e}")

    def sync_frame_detected_callback(self) -> None:
        """Process a sync frame detection."""
        self.__logger.info("Sync frame detected, switching into tracking mode...")
        self.current_mode = self.MODE_DATA
        self.sync_frame_decoder = None
        self.data_frame_decoder = DataFrameDecoder(
            self.__logger,
            self.dataframe_callback,
        )

    def dataframe_callback(
        self,
        good_sync: bool,
        frame_data: DataFrameContents,
    ):
        """Process a data frame."""
        self.__logger.debug(
            (
                "Sensor: {}  TS:{:06x}  Width:{:4x}  "
                + "Chan:{:4}({})  offset:{:-6d}  BeamWord:{:05x}\t"
            ).format(
                frame_data.sid,
                frame_data.timestamp,
                frame_data.width,
                frame_data.base_station_id(),
                frame_data.slow_bit() if frame_data.valid_npoly() else "-",
                frame_data.sync_offset,
                frame_data.beam_word,
            )
        )

        if not good_sync:
            self.__logger.warning("Frame sync lost, switching back to sync mode...")
            self.current_mode = self.MODE_SYNC
            self.sync_frame_decoder = SyncFrameDecoder(
                callback=self.sync_frame_detected_callback,
                logger=self.__logger,
            )
            self.data_frame_decoder = None
        else:
            self.sweep_decoder.process_frame(frame_data)
            # weird handing of the timestamp for the slow bit in
            # the cz handleCalibrationData function
            if (
                frame_data.valid_npoly()
                and (frame_data.base_station_id() < DECK_LIGHTHOUSE_MAX_N_BS)
                and (frame_data.sync_offset != 0)
            ):
                timestamp0 = timestamp_diff(
                    frame_data.timestamp, frame_data.sync_offset
                )
                if self.ootx_decoders[frame_data.base_station_id()] is None:
                    self.__logger.info(
                        "Creating new OOTX decoder for basestation base_station_id {}".format(
                            frame_data.base_station_id()
                        )
                    )
                    self.ootx_decoders[frame_data.base_station_id()] = OOTXFrameDecoder(
                        logger=self.__logger,
                    )
                    self.prev_timestamp0[frame_data.base_station_id()] = timestamp0
                prev_timestamp0 = self.prev_timestamp0[frame_data.base_station_id()]
                if timestamp_abs_diff_larger_than(
                    timestamp0, prev_timestamp0, MIN_TICKS_BETWEEN_SLOW_BITS
                ):
                    self.ootx_decoders[frame_data.base_station_id()].process_slow_bit(
                        frame_data.slow_bit()
                    )
                self.prev_timestamp0[frame_data.base_station_id()] = timestamp0

    def sweep_callback(self, sweep_contents: SweepBlockRawData) -> None:
        """Process a full sweep."""
        self.measurement_processor.process_block(sweep_contents)

    def measurement_callback(self, sensor_bearings: SweepBlockBearings) -> None:
        """Process a measurement."""
        self.__logger.debug(
            "Channel: {base_station_id}".format(
                base_station_id=sensor_bearings.base_station_id
            )
        )
        self.__logger.debug(
            "HW Timestamp: {hardware_timestamp}".format(
                hardware_timestamp=sensor_bearings.hardware_timestamp
            )
        )
        for i in range(PULSE_PROCESSOR_N_SENSORS):
            self.__logger.debug(
                "Sensor {sensor} Azimuth: {azimuth} Elevation: {elevation}".format(
                    sensor=i,
                    azimuth=sensor_bearings.sensor_angles[i].azimuth,
                    elevation=sensor_bearings.sensor_angles[i].elevation,
                )
            )
        if self.__app_bearing_callback is not None:
            self.__app_bearing_callback(sensor_bearings)

File: lighthouse_protocol_decoder/src/lighthouse_protocol_decoder.cpp
// Copyright 2026 Ekumen, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "lighthouse_protocol_decoder/lighthouse_protocol_decoder.hpp"

#include <iomanip>
#include <sstream>

namespace lighthouse_protocol_decoder {

LighthouseProtocolDecoder::LighthouseProtocolDecoder(
    BearingCallback app_bearing_callback, LoggerInterface::Ptr logger)
    : current_mode_(Mode::SYNC),
      app_bearing_callback_(std::move(app_bearing_callback)),
      logger_(std::move(logger)) {

  // Initialize in SYNC mode
  sync_frame_decoder_ = std::make_unique<SyncFrameDecoder>(
      [this]() { syncFrameDetectedCallback(); }, logger_);

  // Create the sweep and measurement processors
  sweep_processor_ = std::make_unique<SweepProcessor>(
      [this](const SweepBlockRawData &sweep) { sweepCallback(sweep); },
      logger_);

  measurement_processor_ = std::make_unique<MeasurementProcessor>(
      [this](const SweepBlockBearings &bearings) {
        measurementCallback(bearings);
      },
      logger_);

  // Initialize timestamp tracking
  prev_timestamp0_.fill(0);
}

void LighthouseProtocolDecoder::processByte(std::uint8_t byte) {
  try {
    if (current_mode_ == Mode::SYNC) {
      if (sync_frame_decoder_) {
        sync_frame_decoder_->processByte(byte);
      }
    } else if (current_mode_ == Mode::DATA) {
      if (data_frame_decoder_) {
        data_frame_decoder_->processByte(byte);
      }
    }
  } catch (const std::exception &e) {
    if (logger_) {
      logger_->error(std::string("Exception caught: ") + e.what());
    }
  }
}

void LighthouseProtocolDecoder::reset() {
  current_mode_ = Mode::SYNC;
  sync_frame_decoder_ = std::make_unique<SyncFrameDecoder>(
      [this]() { syncFrameDetectedCallback(); }, logger_);
  data_frame_decoder_.reset();
  sweep_processor_->reset();
  measurement_processor_->reset();
  std::for_each(ootx_decoders_.begin(), ootx_decoders_.end(),
                [](auto &decoder) { decoder.reset(); });
  std::for_each(prev_timestamp0_.begin(), prev_timestamp0_.end(),
                [](auto &ts) { ts = 0; });
}

void LighthouseProtocolDecoder::syncFrameDetectedCallback() {
  if (logger_) {
    logger_->info("Sync frame detected, switching into tracking mode...");
  }

  // Switch to DATA mode
  current_mode_ = Mode::DATA;
  sync_frame_decoder_.reset();

  // Create data frame decoder
  data_frame_decoder_ = std::make_unique<DataFrameDecoder>(
      [this](bool good_sync, const DataFrameContents &frame) {
        dataframeCallback(good_sync, frame);
      },
      logger_);
}

void LighthouseProtocolDecoder::dataframeCallback(
    bool good_sync, const DataFrameContents &frame_data) {

  if (logger_) {
    std::ostringstream oss;
    oss << "Sensor: " << static_cast<int>(frame_data.sid) << "  TS:" << std::hex
        << std::setw(6) << std::setfill('0') << frame_data.timestamp
        << "  Width:" << std::setw(4) << frame_data.width << std::dec
        << "  Chan:" << std::setw(4)
        << static_cast<int>(frame_data.baseStationId()) << "("
        << (frame_data.validNpoly() ? std::to_string(frame_data.slowBit())
                                    : "-")
        << ")"
        << "  offset:" << std::setw(6) << std::left << frame_data.sync_offset
        << std::right << "  BeamWord:" << std::hex << std::setw(5)
        << std::setfill('0') << frame_data.beam_word << std::dec;
    logger_->debug(oss.str());
  }

  if (!good_sync) {
    if (logger_) {
      logger_->warning("Frame sync lost, switching back to sync mode...");
    }

    // Switch back to SYNC mode
    current_mode_ = Mode::SYNC;
    sync_frame_decoder_ = std::make_unique<SyncFrameDecoder>(
        [this]() { syncFrameDetectedCallback(); }, logger_);
    data_frame_decoder_.reset();
  } else {
    // Process the frame through the sweep processor
    sweep_processor_->processFrame(frame_data);

    // Handle OOTX slow bit processing
    if (frame_data.validNpoly() &&
        frame_data.baseStationId() < kDeckLighthouseMaxNBs &&
        frame_data.sync_offset != 0) {

      const std::uint32_t timestamp0 =
          timestampDiff(frame_data.timestamp, frame_data.sync_offset);

      const std::uint8_t bs_id = frame_data.baseStationId();

      // Create OOTX decoder for this base station if needed
      if (!ootx_decoders_[bs_id]) {
        if (logger_) {
          logger_->info("Creating new OOTX decoder for basestation " +
                        std::to_string(bs_id));
        }
        ootx_decoders_[bs_id] = std::make_unique<OOTXFrameDecoder>(logger_);
        prev_timestamp0_[bs_id] = timestamp0;
      }

      // Filter slow bits based on timing
      const std::uint32_t prev_timestamp0 = prev_timestamp0_[bs_id];
      if (timestampAbsDiffLargerThan(timestamp0, prev_timestamp0,
                                     kMinTicksBetweenSlowBits)) {
        ootx_decoders_[bs_id]->processSlowBit(frame_data.slowBit());
      }

      prev_timestamp0_[bs_id] = timestamp0;
    }
  }
}

void LighthouseProtocolDecoder::sweepCallback(
    const SweepBlockRawData &sweep_contents) {
  // Forward to measurement processor
  measurement_processor_->processBlock(sweep_contents);
}

void LighthouseProtocolDecoder::measurementCallback(
    const SweepBlockBearings &sensor_bearings) {

  if (logger_) {
    logger_->debug("Channel: " +
                   std::to_string(sensor_bearings.base_station_id));
    logger_->debug("HW Timestamp: " +
                   std::to_string(sensor_bearings.hardware_timestamp));

    for (std::size_t i = 0; i < kPulseProcessorNSensors; ++i) {
      std::ostringstream oss;
      oss << "Sensor " << i << " Azimuth: " << std::fixed
          << std::setprecision(2) << sensor_bearings.sensor_angles[i].azimuth
          << " Elevation: " << sensor_bearings.sensor_angles[i].elevation;
      logger_->debug(oss.str());
    }
  }

  // Forward to application callback if provided
  if (app_bearing_callback_) {
    app_bearing_callback_(sensor_bearings);
  }
}

} // namespace lighthouse_protocol_decoder
